cuda_core/cuda/core/experimental/_device.pyx

```diff
@@ -1145,6 +1145,49 @@ class Device:
     def __repr__(self):
         return f"<Device {self._id} ({self.name})>"
 
+    def __hash__(self) -> int:
+        """Return hash based on the device ordinal.
+
+        This enables Device objects to be used as dictionary keys and in sets.
+        Device objects with the same device_id will hash to the same value
+        and be considered equal, even if they are different Python objects or
+        exist on different threads.
+
+        Returns
+        -------
+        int
+            Hash value based on the device ordinal (device_id).
+
+        Notes
+        -----
+        Device is a per-thread singleton, but equality is based on logical
+        device identity (device_id), not Python object identity. This means
+        Device(0) on thread A equals Device(0) on thread B.
+        """
+        return hash(self._id)
```

Comment:
Since this is the device ordinal, it can in theory change across processes if someone uses something like `CUDA_VISIBLE_DEVICES`. Should we consider hashing based on something like the UUID that remains consistent regardless of how the devices were ordered?

Reply:
Yikes. I didn't know about `CUDA_VISIBLE_DEVICES` -- seems like a dangerous feature.  I definitely think there is a value in utilizing a UUID that is guaranteed to uniquely identify a device across processes.  It would be great if we had a more practical use case b/c the `CUDA_VISIBLE_DEVICES` example comes across as a user intentionally trying to break the system.  That being said I'll move forward with making the change.

I ran some benchmark experiments and using an UUID in the hash has minimal impact on performance provided we cache the UUID in the device object. That enables us to avoid calling a CUDA API per hash request which is 10x slower than the hash calculation.

  • UUID fetch (CUDA API call): 0.35–0.37 microseconds
  • Hash computation (tuple hash): 0.03–0.04 microseconds

Reply:
Done.